pkg/server/tls_windows.go

```diff
@@ -0,0 +1,342 @@
+		return nil, err
+	}
+	filtered, err = c.filterBySubjectRegularExpression(filtered, subjectRegEx)
+	if err != nil {
+		return nil, err
+	}
+	if len(filtered) == 0 {
+		return nil, fmt.Errorf("no certificates found")
+	}
+	// order oldest to newest
+	sort.Slice(filtered, func(i, j int) bool {
+		// Already accessed this so the error will not happen
+		a, _ := filtered[i].Certificate()
+		b, _ := filtered[j].Certificate()
+
+		return a.NotBefore.Before(b.NotBefore)
+	})
+
+	// Grab the most recent valid one
+	for i := len(filtered) - 1; i >= 0; i-- {
+		fc, _ := filtered[i].Certificate()
+		if time.Now().Before(fc.NotAfter) && time.Now().After(fc.NotBefore) {
+			validIdentity = filtered[i]
+			break
+		}
+	}
+	if validIdentity == nil {
+		return nil, fmt.Errorf("no certificates found")
+	}
+
+	return validIdentity, nil
+}
+
+func (c *winCertStoreHandler) filterByIssuerCommonNames(input []certstore.Identity, commonNames []string) ([]certstore.Identity, error) {
+	if len(commonNames) == 0 {
+		return input, nil
+	}
+	returnIdentities := make([]certstore.Identity, 0)
+	for _, identity := range input {
+		cert, err := identity.Certificate()
+		if err != nil {
+			return nil, err
+		}
+		for _, cfgName := range commonNames {
+			if cert.Issuer.CommonName == cfgName {
+				returnIdentities = append(returnIdentities, identity)
+				break
+			}
+		}
+	}
+	return returnIdentities, nil
+}
+
+func (c *winCertStoreHandler) filterByTemplateID(input []certstore.Identity, id string) ([]certstore.Identity, error) {
+	if id == "" {
+		return input, nil
+	}
+	returnIdentities := make([]certstore.Identity, 0)
+
+	for _, identity := range input {
+		cert, err := identity.Certificate()
+		if err != nil {
+			return nil, err
+		}
+		for _, ext := range cert.Extensions {
+			if ext.Id.String() == asnTemplateOID {
+				templateInfo := &templateInformation{}
+				_, err := asn1.Unmarshal(ext.Value, templateInfo)
+				if err != nil {
+					return nil, err
+				}
+				if templateInfo.Template.String() == id {
+					returnIdentities = append(returnIdentities, identity)
+				}
+			}
+		}
+	}
+	return returnIdentities, nil
+}
+
+func (c *winCertStoreHandler) filterBySubjectRegularExpression(input []certstore.Identity, regEx *regexp.Regexp) ([]certstore.Identity, error) {
+	if regEx == nil {
+		return input, nil
+	}
+	returnIdentities := make([]certstore.Identity, 0)
+
+	for _, identity := range input {
+		cert, err := identity.Certificate()
+		if err != nil {
+			return nil, err
+		}
+		if regEx.MatchString(cert.Subject.CommonName) {
+			returnIdentities = append(returnIdentities, identity)
+		}
+	}
+	return returnIdentities, nil
+}
+
+func (c *winCertStoreHandler) VerifyPeer(_ [][]byte, verifiedChains [][]*x509.Certificate) error {
+	opts := x509.VerifyOptions{}
+	clientCert := verifiedChains[0][0]
+	_, err := clientCert.Verify(opts)
+	return err
+
+}
+
+type templateInformation struct {
+	Template     asn1.ObjectIdentifier
+	MajorVersion int
+	MinorVersion int
+}
```

Comment:
nit: can this be declared directly after filterByTemplateID? I was thrown off by what was using this on first glance.